src/core/job.c

```diff
@@ -962,6 +968,10 @@ int job_finish_and_invalidate(Job *j, JobResult result, bool recursive, bool alr
 
         if (IN_SET(result, JOB_FAILED, JOB_INVALID))
                 j->manager->n_failed_jobs++;
+        else if (result == JOB_DONE && u->activation_event_info == NULL && IN_SET(t, JOB_START, JOB_RESTART))
+                u->activation_event_info = TAKE_PTR(j->activation_event_info); /* Older event wins */
+        else if (result == JOB_DONE && t == JOB_STOP)
+                u->activation_event_info = activation_event_info_unref(u->activation_event_info);
```

Comment:
Hmm, so you are filling this in here after activation is complete, and invalidate it after deactivation is complete?

I think this isn#t right. I am sure this should be installed whenever we start activation, i.e. the `.start()` vtable functions of the unit types should be able to look in there, if they like.

Hence, I think unit_start() should take a pointer to the new struct as (optional) parameter, and install it already before invoking that vtable function `.start()`.

Moreover, I think we should leave the object in place when we stop the thing, so that (unless a unit is GC'ed) we can still see the info all the way until the next activation.

Or with other words, this info should have the life-time of most runtime bits we have, such as the invocation ID: we reset/initialize it in unit_start(), and then leave it around until basically the next unit_start().